### PR TITLE
`Availability` concern

### PR DESCRIPTION
**Pull request by davegudge**
Introduce an `Availability` concern to encapsulate the logic to determine if a developer is available by utilising an `enum`.

This PR [originally](88cd5b242fdb991044cd41776c00e0461907130a) DRY'ed up the view logic by utilising a partial, however, the logic in #59 has been cherry-picked and adapted to work with the `Availability` concern.

*N.B:* This PR retains the previous styling / [wording](https://github.com/joemasilotti/railsdevs.io/pull/58/files#diff-44438ce218f5287c58d0017f965d888715635d94280669896f75841fbd7b4cd7R17-R20). i.e there are no changes to the UI from a user perspective.

---
Screenshot 2021-11-07 at 13 59 34

Screenshot 2021-11-07 at 13 59 27

### CODE REVIEW

File: app/views/developers/_availability.html.erb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<span class="text-gray-900 text-sm <%= "font-medium" if developer.available_now? %>">
```
Comment (joemasilotti) [inline app/views/developers/_availability.html.erb:1]: This file can be removed, right?

File: app/components/availability_component.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
```
Comment (joemasilotti) [inline app/components/availability_component.rb:1]: Please remove this comment.

File: app/models/concerns/availability.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
```
Comment (joemasilotti) [inline app/models/concerns/availability.rb:1]: Please remove this comment.

File: test/models/developer_test.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
```
Comment (joemasilotti) [inline test/models/developer_test.rb:1]: Please remove this comment.

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This looks amazing! I love the architecture of the code.

Unfortunately, I don't think the concern is working as expected. `#available_on=` doesn't seem to be getting called when the record is loaded from the database. So if you visit the index view you can see that all developer's are "Currently unavailable" even though the seeds data has 2 available devs.

File: app/views/developers/_availability.html.erb (hunk 1 above)
Comment (davegudge) [reply app/views/developers/_availability.html.erb:1]: Good spot! I must have forgotten to delete this file when cherry-picking the component commit. I guess this is why we code review 😉

File: app/components/availability_component.rb (hunk 2 above)
Comment (davegudge) [reply app/components/availability_component.rb:1]: No problem. 

I was going to suggest that this magic comment improves performance (https://www.mikeperham.com/2018/02/28/ruby-optimization-with-one-magic-comment/), but that does not always appear to be the case according to this post: https://accidentaltechnologist.com/ruby-on-rails/frozen_string_literal-the-not-so-magical-comment/

Comment (joemasilotti) [reply app/components/availability_component.rb:1]: I seem to remember that it's not needed with Ruby 3.0 either.

File: app/models/concerns/availability.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+module Availability
+  extend ActiveSupport::Concern
+
+  included do
+    enum availability_status: [:unspecified, :now, :in_future], _default: :unspecified, _prefix: :available
```
[the remark this answers is not in the document]
Comment (davegudge) [reply app/models/concerns/availability.rb:7]: This `enum` is not backed by a db column at the moment, but that may change when #46 is actioned.

**Comment by davegudge** [comment]
New commits pushed onto this PR.

**Review by joemasilotti — APPROVED** [review]
Looks great, thanks for the fixes!

**Pull request merged**